workflow/scripts/add_district_heating_subnodes.py

```diff
@@ -0,0 +1,421 @@
+
+    # Convert the DataFrame to a GeoDataFrame
+    subnodes = gpd.GeoDataFrame(subnodes, crs="EPSG:4326")
+
+    # Assign cluster to subnodes according to onshore regions
+    subnodes["cluster"] = subnodes.apply(
+        lambda x: regions_onshore.geometry.contains(x.geometry).idxmax(), axis=1
+    )
+    subnodes["lau"] = subnodes.apply(
+        lambda x: lau.loc[lau.geometry.contains(x.geometry).idxmax(), "LAU_ID"], axis=1
+    )
+    subnodes["lau_shape"] = subnodes.apply(
+        lambda x: lau.loc[lau.geometry.contains(x.geometry).idxmax(), "geometry"].wkt,
+        axis=1,
+    )
+
+    return subnodes
+
+
+def add_subnodes(n: pypsa.Network, subnodes: gpd.GeoDataFrame) -> None:
+    """
+    Add largest district heating systems subnodes to the network. They are initialized with
+     - the total annual heat demand taken from the mother node, that is assigned to urban central heat and low-temperature heat for industry,
+     - the heat demand profiles taken from the mother node,
+     - and the district heating investment options (stores, links) from the mother node,
+     - and heat vents as generator components
+    The district heating loads in the mother nodes are recuded accordingly.
+
+    Parameters:
+    n (pypsa.Network): The PyPSA network object to which subnodes will be added.
+    subnodes (gpd.GeoDataFrame): GeoDataFrame containing information about district heating subnodes.
+
+    Returns:
+    None
+    """
+
+    # Add subnodes to network
+    for _, subnode in subnodes.iterrows():
+        name = f'{subnode["cluster"]} {subnode["Stadt"]} urban central heat'
+
+        # Add buses
+        n.madd(
+            "Bus",
+            [name],
+            y=subnode.geometry.y,
+            x=subnode.geometry.x,
+            country="DE",
+            location=f"{subnode['cluster']} {subnode['Stadt']}",
+            carrier="urban central heat",
+            unit="MWh_th",
+        )
+
+        # Add heat loads for urban central heat and low-temperature heat for industry
+        uch_load_cluster = (
+            n.snapshot_weightings.generators
+            @ n.loads_t.p_set[f"{subnode['cluster']} urban central heat"]
+        )
+        lti_load_cluster = (
+            n.loads.loc[
+                f"{subnode['cluster']} low-temperature heat for industry", "p_set"
+            ]
+            * 8760
+        )
+        dh_load_cluster = uch_load_cluster + lti_load_cluster
+        lti_share = lti_load_cluster / dh_load_cluster
+
+        demand_ratio = min(
+            1,
+            (subnode["yearly_heat_demand_MWh"] / dh_load_cluster),
+        )
+
+        lost_load = subnode["yearly_heat_demand_MWh"] - dh_load_cluster
+
+        if demand_ratio == 1:
+            logger.info(
+                f"District heating load of {subnode['Stadt']} exceeds load of its assigned cluster {subnode['cluster']}. {lost_load} MWh/a are disregarded."
+            )
+        uch_load = (
```

Comment:
I think, a few comments on lines like this and similar operations would help the readability.